### PR TITLE
kubelet: set kubelet parameters via an intermediate configuration file

### PR DESCRIPTION
**Pull request by cafkafk**
## Description of changes
Kubelets have deprecated using arguments such as `--address`, and we should
instead use the `KubeletConfiguration` struct as specified in
https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/#kubelet-conf-d.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by cafkafk** [comment]
It *seems* like the only deprecation warning for the kubelet now is:
```
machine1 # [    9.025319] kubelet[1116]: Flag --pod-infra-container-image has been deprecated, will be removed in a future release. Image garbage collector will get sandbox image information from CRI.
```

**Review by vdemeester — APPROVED** [review]
LGTM 🐯

File: nixos/modules/services/cluster/kubernetes/kubelet.nix [hunk 1]
```diff
@@ -33,6 +33,41 @@ let
 
   kubeconfig = top.lib.mkKubeConfig "kubelet" cfg.kubeconfig;
 
+  # Flag based settings are deprecated, use the `--config` flag with a
+  # `KubeletConfiguration` struct.
+  # https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/
+  #
+  # NOTE: registerWithTains requires a []core/v1.Taint, therefore requires
```
Comment (jonpulsifer) [inline nixos/modules/services/cluster/kubernetes/kubelet.nix:40]:
```suggestion
  # NOTE: registerWithTaints requires a []core/v1.Taint, therefore requires
```

Comment (cafkafk) [reply nixos/modules/services/cluster/kubernetes/kubelet.nix]: Nice catch :+1: fixed in force push

File: nixos/modules/services/cluster/kubernetes/kubelet.nix [hunk 2]
```diff
@@ -33,6 +33,41 @@ let
...
+  # NOTE: registerWithTaints requires a []core/v1.Taint, therefore requires
+  # additional work to be put in config format.
+  #
+  kubeletConfig = pkgs.writeText "kubelet-config" (builtins.toJSON ({
+    apiVersion = "kubelet.config.k8s.io/v1beta1";
+    kind = "KubeletConfiguration";
+    address = cfg.address;
+    port = cfg.port;
+    authentication = {
+      x509 = { } // (lib.optionalAttrs (cfg.clientCaFile != null) { clientCAFile = cfg.clientCaFile; });
```
Comment (SuperSandro2000) [inline nixos/modules/services/cluster/kubernetes/kubelet.nix:49]:
```suggestion
      x509 = lib.optionalAttrs (cfg.clientCaFile != null) { clientCAFile = cfg.clientCaFile; };
```
optionalAttrs returns an empty attr on false condition

File: nixos/modules/services/cluster/kubernetes/kubelet.nix [hunk 3]
```diff
@@ -33,6 +33,41 @@ let
...
+    registerNode = cfg.registerNode;
+    containerRuntimeEndpoint = cfg.containerRuntimeEndpoint;
+    healthzPort = cfg.healthz.port;
+    healthzBindAddress = cfg.healthz.bind;
+  } // (lib.optionalAttrs (cfg.tlsCertFile != null)  { tlsCertFile = cfg.tlsCertFile; })
+    // (lib.optionalAttrs (cfg.tlsKeyFile != null)   { tlsPrivateKeyFile = cfg.tlsKeyFile; })
+    // (lib.optionalAttrs (cfg.clusterDomain != "")  { clusterDomain = cfg.clusterDomain; })
+    // (lib.optionalAttrs (cfg.clusterDns != "")     { clusterDNS = [ cfg.clusterDns ] ; })
+    // (lib.optionalAttrs (cfg.featureGates != [])   { featureGates = cfg.featureGates; })
+  ));
```
Comment (SuperSandro2000) [inline nixos/modules/services/cluster/kubernetes/kubelet.nix:69]:
```suggestion
  } // lib.optionalAttrs (cfg.tlsCertFile != null)  { tlsCertFile = cfg.tlsCertFile; }
    // lib.optionalAttrs (cfg.tlsKeyFile != null)   { tlsPrivateKeyFile = cfg.tlsKeyFile; }
    // lib.optionalAttrs (cfg.clusterDomain != "")  { clusterDomain = cfg.clusterDomain; }
    // lib.optionalAttrs (cfg.clusterDns != "")     { clusterDNS = [ cfg.clusterDns ] ; }
    // lib.optionalAttrs (cfg.featureGates != [])   { featureGates = cfg.featureGates; }
  );
```
those are purely optional

File: nixos/modules/services/cluster/kubernetes/kubelet.nix (hunk 2 above)
Comment (cafkafk) [reply nixos/modules/services/cluster/kubernetes/kubelet.nix:49]: I see, will fix :+1:

File: nixos/modules/services/cluster/kubernetes/kubelet.nix (hunk 3 above)
Comment (cafkafk) [reply nixos/modules/services/cluster/kubernetes/kubelet.nix]: Also fixing, ty :)

**Pull request merged**